### PR TITLE
Create XML Merge function

### PR DESCRIPTION
**Pull request by LucasRochaAbraao**
Create a function that edits a junit xml produced by pytest and updates test results in the original file.

EDEN-2543

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
+from typing import Dict
+
+from _pytest.reports import TestReport
+
+from src.pytest_merge_xml.merge_xml import modify_xml
```
Comment (igortg) [inline conftest.py]: This should be:
```suggestion
from pytest_merge_xml.merge_xml import modify_xml
```

File: conftest.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    parser.addoption(
+        "--mergexml",
+        action="store",
+        help="Enable original xml overwrite based on flaky tests rerun results",
+    )
+
+
+def pytest_configure(config):
+    original_xmlfile = config.getoption("--mergexml")
+    # report = XMLReport(Path('src/pytest_merge_xml/failure.xml'))
```
Comment (raispbizerra) [inline conftest.py]: Left-over?

File: src/pytest_merge_xml/merge_xml.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    root = tree.getroot()
+    testsuite = root.find("testsuite")
+    failure_count = int(testsuite.attrib['failures'])
+
+    results_passed = [
+        result.location[2] for result in retest_results.values() if result.outcome == 'passed'
+    ]
+
+    for testcase in testsuite:
+        testcase_name = testcase.attrib['name']  # 'test_xml_one_failed_two_passed'
```
Comment (raispbizerra) [inline src/pytest_merge_xml/merge_xml.py]: What's this comment about?

File: pyptoject.toml [hunk 4]
```diff
@@ -1,7 +1,8 @@
 [tools.isort]
```
Comment (nicoddemus) [inline pyptoject.toml]: The name of the file is wrong, should be `pyproject.toml`

File: src/pytest_merge_xml/merge_xml.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        for failure in testcase.findall("failure"):
+            print(testcase_name, results_passed)
```
Comment (igortg) [inline src/pytest_merge_xml/merge_xml.py]: Remove this `print` statement.

File: src/pytest_merge_xml/merge_xml.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+from typing import Dict
+
+from _pytest.reports import TestReport
+
+
+def modify_xml(original_xml: Path, retest_results: Dict[str, TestReport], new_xml: Path) -> None:
+    tree = Et.parse(original_xml)
+    root = tree.getroot()
+    testsuite = root.find("testsuite")
+    failure_count = int(testsuite.attrib['failures'])
```
Comment (igortg) [inline src/pytest_merge_xml/merge_xml.py]:
```suggestion
    previous_run_failure_count = int(testsuite.attrib['failures'])
```

File: .gitignore [hunk 7]
```diff
@@ -18,3 +18,7 @@ dist/
 .eggs/
 .tox/
 .venv/
+
+# local files
```
Comment (igortg) [inline .gitignore]: Do not ignore thoses.

File: conftest.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
+from typing import Dict
+
+from _pytest.reports import TestReport
+
+from src.pytest_merge_xml.merge_xml import modify_xml
+
+
+class XMLReport:
```
Comment (igortg) [inline conftest.py]: Add some docstring:
Add a docstring. Something like:
```suggestion
class XMLReport:
    '''Class that collect test items and update and existing XML file with the new test results'''
    
```

**Review by igortg — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -9,6 +9,6 @@ repos:
   - id: flake8
 - repo: https://github.com/pycqa/isort
   rev: 5.10.1
-    hooks:
-    - id: isort
-      name: isort (python)
+  hooks:
+  - id: isort
+    name: isort (python)
```
Comment (nicoddemus) [inline .pre-commit-config.yaml]: Missing `black` for formatting

File: conftest.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
```
Comment (nicoddemus) [inline conftest.py]: If this is meant to be a plugin installed by others, you cannot use a `conftest.py` file for this, you make this a normal module and use [setuptools entry points to register it](https://docs.pytest.org/en/stable/how-to/writing_plugins.html#making-your-plugin-installable-by-others).

File: conftest.py (hunk 2 above)
Comment (LucasRochaAbraao) [reply conftest.py]: Yeah, I'll remove it.

File: src/pytest_merge_xml/merge_xml.py (hunk 3 above)
Comment (LucasRochaAbraao) [reply src/pytest_merge_xml/merge_xml.py]: I was using it as reference during development, I'll remove it as well!

File: pyptoject.toml (hunk 4 above)
Comment (LucasRochaAbraao) [reply pyptoject.toml]: Thanks 👍

File: conftest.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    def pytest_sessionfinish(self, session, exitstatus):
+        modify_xml(Path(self.original_xmlfile), self.reports, Path(f"{self.original_xmlfile}.ovrw"))
+
+
+def pytest_addoption(parser):
+    # pytest has a note to keep conftest.py at the root dir in order to load added options.
+    # http://pytest.org/latest/plugins.html?highlight=hooks#plugin-discovery-order-at-tool-startup
+    parser.addoption(
+        "--mergexml",
```
Comment (nicoddemus) [inline conftest.py]: I find it a bit confusing that the plugin is named `pytest-merge-xml` and the flag is `--mergexml`, I suggest:

```suggestion
        "--merge-xml",
```

File: src/pytest_merge_xml/merge_xml.py (hunk 5 above)
Comment (LucasRochaAbraao) [reply src/pytest_merge_xml/merge_xml.py]: Removed the statement.

File: src/pytest_merge_xml/merge_xml.py (hunk 6 above)
Comment (LucasRochaAbraao) [reply src/pytest_merge_xml/merge_xml.py]: Renamed the variable.

File: .gitignore (hunk 7 above)
Comment (LucasRochaAbraao) [reply .gitignore]: Fixed it, not ignoring the failure.xml file anymore. The other one isn't needed, I was using it to keep the original unchanged so I didn't have to recreate it.

File: conftest.py (hunk 8 above)
Comment (LucasRochaAbraao) [reply conftest.py]: Done!

File: conftest.py (hunk 10 above)
Comment (LucasRochaAbraao) [reply conftest.py]: Thanks for the correction! I changed it and it's working now. I installed the package locally and tested in a fake project to make sure.

File: .pre-commit-config.yaml (hunk 9 above)
Comment (LucasRochaAbraao) [reply .pre-commit-config.yaml]: It's now included, thanks!

File: conftest.py (hunk 11 above)
Comment (LucasRochaAbraao) [reply conftest.py]: Makes sense, thanks for the suggestion! I changed it.

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -9,6 +9,11 @@ repos:
   - id: flake8
 - repo: https://github.com/pycqa/isort
   rev: 5.10.1
-    hooks:
-    - id: isort
-      name: isort (python)
+  hooks:
+  - id: isort
```
Comment (nicoddemus) [inline .pre-commit-config.yaml]: @igortg do you feel strongly about `isort`? I personally have been using [reorder-python-imports](https://pypi.org/project/reorder-python-imports/) in all my repos.

File: setup.cfg [hunk 13]
```diff
@@ -3,3 +3,4 @@ universal = 1
 
 [flake8]
 exclude = docs
+max-line-length = 100
```
Comment (nicoddemus) [inline setup.cfg:6]: Add a `mypy` session here, using this one as basis: https://github.com/ESSS/pytest-regressions/blob/master/mypy.ini

Make sure to include `disallow_untyped_defs = True` too, to ensure all functions/methods are all type-annotated.

File: src/pytest_merge_xml/plugin.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
+from pathlib import Path
+from typing import Dict
+
+from _pytest.reports import TestReport
+
+from pytest_merge_xml.merge_xml import modify_xml
+
+
+class XMLReport:
```
Comment (nicoddemus) [inline src/pytest_merge_xml/plugin.py]:
```suggestion
class XMLReportPlugin:
```

Also XMLReport does not really conveys what the class means, how about `XMLReportMergerPlugin`, or something like that?

File: test/test_pytest_merge_xml.py [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        ),
+    }
+    obtained_file = datadir / "failure_tmp.xml"
+    modify_xml(
+        Path(datadir / "failure.xml"), retest_results_some_passing, obtained_file
+    )
+    file_regression.check(obtained_file.read_text(), extension=".xml")
+
+
+def test_xml_three_failed_zero_passed(
```
Comment (nicoddemus) [inline test/test_pytest_merge_xml.py:109]: Those tests are all valid, but I suggest at least one integration test using the `pytester` fixture.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
I think we are almost there, please take a look at my comments. :+1:

File: src/pytest_merge_xml/plugin.py (hunk 14 above)
Comment (LucasRochaAbraao) [reply src/pytest_merge_xml/plugin.py]: In terms of conveying what the class means, I'm leaning towards 'XMLReportMergerPlugin'. 'XMLReportPlugin' also feels a bit generic, and this class actually modifies the resulting xml. Thanks for the suggestion!

**Comment by igortg** [comment]
Right now I'd like to redirect all the effort to validate that this plugin would really work on our environment (being called within `inv test-py`).

I created [EDEN-2553](https://eden.esss.co/jira/browse/EDEN-2553) and [EDEN-2554](https://eden.esss.co/jira/browse/EDEN-2554) so we can tackle @nicoddemus suggestions as soon as we make make this validation.

File: .pre-commit-config.yaml (hunk 12 above)
Comment (igortg) [reply .pre-commit-config.yaml]: Better... @LucasRochaAbraao will do the change.

**Review by igortg — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 12 above)
Comment (LucasRochaAbraao) [reply .pre-commit-config.yaml]: Changed it.

File: test/test_pytest_merge_xml.py (hunk 15 above)
Comment (LucasRochaAbraao) [reply test/test_pytest_merge_xml.py:109]: I'll look into this in a later issue! Thanks.

File: setup.cfg (hunk 13 above)
Comment (LucasRochaAbraao) [reply setup.cfg:6]: I'll look into this in a later issue!

**Pull request merged**